design/002-secret-generator-group.md

```diff
@@ -0,0 +1,277 @@
+of services running inside clusters (i.e. databases etc).
+
+### Goals
+
+* enable zero-trust secret management for purely technical and internal users
+* enable secure passwords by default
+* reduce error rates when creating RSA or ECDSA private keys
+
+### Non-Goals
+
+* auto-rotation of secrets
+
+## Proposal
+
+Implement several generators to support not only a number of data types but also generators that interface with external APIs.
+
+### API
+
+We add a new resource group `generator.external-secrets.io`. For each generator we add a new CRD, e.g. for password generation, VaultPKI, ECR, GCR etc.
+This custom resource contains all the configuration details to create a secure value.
+It can be referenced from a `Kind=ExternalSecret` or `Kind=PushSecret`. **or embedded** into it.
+
+The controller generates the value and passes it onto the `Kind=Secret` with all the values
+that have been fetched from the provider.
+When used with a `Kind=PushSecret` it will generate the value and push it to the configured provider.
+
+A generator generates a known set of outputs, e.g. `username` and `password`, `private_key` and `certificate` or others.
+
+#### Example CRD
+
+The following custom resource `Kind=VaultPKIBackend` contains all the parameters for issuing a certificate via HashiCorp Vault PKI engine.
+
+```yaml
+apiVersion: generator.external-secrets.io/v1alpha1
+kind: VaultPKIBackend
+metadata:
+  name: my-vault-pki-creds
+spec:
+  # specific params for generating
+  commonName: example.com
+  altNames:
+  - example.org
+  - localhost
+  # ... more opts
+  auth: {} # service account etc. etc.
+```
+
+This resource can be **embedded or referenced** from an ExternalSecret or PushSecret.
+The ExternalSecret controller generates the secret value and it can be post-processed with
+existing features like template or key-rewrite.
+
+```yaml
+apiVersion: external-secrets.io/v1beta1
+kind: ExternalSecret
+metadata:
+  name: example
+spec:
+  # ...
+  dataFrom:
+
+    # embedded
+    - generator:
+        apiVersion: generator.external-secrets.io/v1alpha1
+        kind: VaultPKIBackend
+        spec:
+          # same as above
+          commonName: example.com
+          altNames:
+          - example.org
+          - localhost
+          # ... more opts
+          auth: {} # service account etc. etc.issuerRef: "default"
+
+    # the generator also can be referenced like here
+    - generatorRef:
+        apiVersion: generator.external-secrets.io/v1alpha1
+        kind: VaultPKIBackend
+        name: foobar
+```
+
+
+<details>
+<summary>PushSecret example</summary>
+
+```yaml
+apiVersion: external-secrets.io/v1alpha1
+kind: PushSecret
+metadata:
+  name: "hello-world"
+spec:
+  secretStoreRefs: [] # ...
+  selector:
+    # we can consider supporting multiple generators per PushSecret
+    # but need to take care of key collision
+    generatorRef:
+      apiVersion: generator.external-secrets.io/v1alpha1
+      kind: VaultPKIBackend
+      name: foobar
```

Comment:
The idea would also allow the `generator` as well as `generatorRef`, right?

Reply:
yeah, i think both options would be needed.
1. inline `generator` for specific per-external-secret configuration
2. `generatorRef` that points to a reusable generator config (e.g. Password/Bytes/Hex/...)